### PR TITLE
[stdlib] Workaround to fix getSectionInfo() for Android

### PR DESCRIPTION
**Pull request by johnno1962**
As discussed in https://github.com/apple/swift/pull/9869, this is a temporary patch to allow stdlib to run on Android. The function getSectionInfo() in stdlib/public/runtime/ImageInspectionELF.cpp tries to dlopen() an image for which the full path is not available on Android causing a fatalError(). This PR patches this out for now, leaving a warning during compilation to indicating it should be revisted at a later date. No Swift program can run on Android without this patch.

Workaround for [SR-5414](https://bugs.swift.org/browse/SR-5414).

### CODE REVIEW

**Comment by gparker42** [comment]
I don't see any compile-time warning, but that's okay because we generally don't use warnings to track todo items. 

There should be an open bug tracking the future fix, and the comment should refer to that bug. John, do you want to file a new bug for the future fix, or do you want to keep SR-5414 open and repurpose it?

**Comment by johnno1962** [comment]
It will give a warning “code never executed” for the fataError() on Android. How should I repurpose SR-5414 - did I set it up wrong? I dont know of a fix for this at present as android goes out of it’s way not to reveal the path to executables so the dlopen() (to inspect the image for protocol conformance?? etc.) will never work. I don’t know what the ramifications of this are myself. Things seem to work!

**Comment by johnno1962** [comment]
Perhpas my use of the words “Temporary” and “resolves” was in error. More like

**Comment by gparker42** [comment]
Describing this change as a workaround for SR-5414 would be fine. 

I suspect that simply giving up will break some protocol conformance checks and reflection. Are you able to run Swift's tests on Android? I hope that some of our tests are able to catch any missing registration here.

**Comment by johnno1962** [comment]
I’ve never run the full test suite on Android myself though I believe it is possible. https://github.com/apple/swift/blob/master/docs/Android.md. The only known problem with Swift itself  on Android I’ve encountered is that try/catch SEGFAULT’s out if you try to catch the error in a var.

**Comment by johnno1962** [comment]
I ran the tests but it’s not performing them on the device so there isn’t much to learn there.

**Comment by johnno1962** [comment]
I’ve created a new JIRA for the do/try/catch problems https://bugs.swift.org/browse/SR-5438. I’m running a Swift from about a month ago but I don’t think this is the problem. I’ll update the Jira if this goes away when I sync to master as swift4 comes close to release.

**Comment by johnno1962** [comment]
Hi @gparker42, I’ve researched this problem a little more and updated the underlying Jira’s if you want more detail. I had thought the two Jiras discussed could be related somehow. After adding some logging seems that the dlopen() is successful even though the full path is supplied with the exception of one outlier which seems to be specific to Android. This means this PR is valid as is, simply making a failure to dlopen non-fatal for Android.

**Comment by johnno1962** [comment]
I’ve done some more logging and I’m satisfied that protocol conformances are being imported correctly on Android so I don’t have an explanation why do/try/catch isn’t working at all.

**Comment by johnno1962** [comment]
I’ve removed the compilation warning as I’m pretty sure this has been investigated on Android now. I’ve also synced my swift sources to master and everything apart from do/try/catch is working which I’m pretty sure is a separate issue.

File: stdlib/public/runtime/ProtocolConformance.cpp [hunk 1]
```diff
@@ -28,13 +28,19 @@
...
 #include <objc/runtime.h>
 
 static const char *class_getName(const ClassMetadata* type) {
   return class_getName(
     reinterpret_cast<Class>(const_cast<ClassMetadata*>(type)));
 }
+#else
+static const char *class_getName(const ClassMetadata* type) {
+  return type->getNominalTypeDescriptor()->Name.get();
+}
```
Comment (jckarter) [inline stdlib/public/runtime/ProtocolConformance.cpp]: Why is this necessary? We shouldn't be using ObjC functions if ObjC interop isn't enabled.

Comment (johnno1962) [inline stdlib/public/runtime/ProtocolConformance.cpp]: Seemed like the least intrusive change to get this code working on non-ObjC interop platforms so I could debug the protocol compliance import from ELF. It’s unfortunate it uses a functoin name from the runtime I supppose.

Comment (johnno1962) [inline stdlib/public/runtime/ProtocolConformance.cpp]: Reverted. It’s not important to this PR.

Comment (jckarter) [inline stdlib/public/runtime/ProtocolConformance.cpp]: No, using getNominalTypeDescriptor() is fine, but there's no reason to wrap it in something that looks like and could be confused for the ObjC runtime function name.

Comment (johnno1962) [inline stdlib/public/runtime/ProtocolConformance.cpp]: Fair enough. I’ve reverted it all anyway. Let me know if you’d like a version that doesn’t use the name class_getName back. It was handy on Linux.

Comment (jckarter) [inline stdlib/public/runtime/ProtocolConformance.cpp]: `swift_getTypeName` should do what you want already more generally.

**Comment by johnno1962** [comment]
Chances of a squish/merge? This is such a small change!

**Comment by gonzalolarralde** [comment]
The reason why this is required for bionic is because `dl_iterate_phdr` is not only returning the currently opened libraries, but also the executable file. When the iterator reaches the executable file path, `RTLD_NOLOAD` avoid loading it at all, so the return value for that call is `null`.

This is what glibc's `dl_iterate_phdr` would return for `exec_test`:

1: `null` < will return a reference to the originating executable who started the process when `dlopen` is called 
2: `loadedlib1.so`
3: `loadedlib2.so`

Vs bionic's `dl_iterate_phdr`:

1: `null` < same as glibc, `dlopen` will get a referece to itself
2: `exec_test` < this won't count as an opened lib so `RTLD_NOLOAD` will make `dlopen` to ignore it
3: `loadedlib1.so`
4: `loadedlib2.so`

# Potential solutions to this problem

### Get the executable name and ignore that entry specifically.

I couldn't find a reliable way to get this. There are some SO responses suggesting calling `dl_iterate_phdr` and taking the first result, but  I couldn't find documentation sustaining this. There may be even problems when this is running in a java environment.

### Have a static variable that keeps the iteration repetition value.

But if this is called more than once then it fails. I don't think it is tho.

### Ignoring any non-loaded dlopen.

The assumption here is that if something is returned by `dl_iterate_phdr` then it should be loaded, on bionic the executable is clearly the exception. It _is_ loaded, you can get the reference by calling `dlopen(null…`, but the actual path that it is passing is not in the table of loaded libraries, so the call returns empty.

If the library is loaded, the conformances processed and then unloaded again, when the conformances are checked the memory will be deallocated and it's going to crash.

# So...

Ignoring non loaded libraries sound like the safe path to me, because even if there's a conformance to be declared it will fail as soon as you want to access it. But we can totally be missing something.

@gparker42 @jckarter please let us know if any of this solutions make sense or if there's any other better alternative that we should consider.

Thanks!

**Comment by jckarter** [comment]
Ignoring libraries that don’t get loaded seems fine.

**Comment by johnno1962** [comment]
OK, I’ve succeeded in squashing the commits.

**Comment by johnno1962** [comment]
Opened a new PR on swift-llvm https://github.com/apple/swift-llvm/pull/55 to fix the do/try/catch problems I mentioned before. Details are in the updated Jira: https://bugs.swift.org/browse/SR-5438

**Comment by johnno1962** [comment]
HI @gparker42, Any change of getting this very minor change merged?

**Comment by johnno1962** [comment]
Closed in favour of https://github.com/apple/swift/pull/11615. Hallelujah!
